<a href="https://colab.research.google.com/github/omar-tueme/ciencia-de-datos-avanzada-mek-enero-2026/blob/main/01_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 3.2 Limpieza y Preparación
### Adriana Ale Huesca - A01022013
### Margarita Ojeda Hernández - A01797688
### Omar Jacobo Tueme Márquez - A00959999


## Preparativos y análisis general de la base de datos

Antes de empezar, vamos a importar las librerias que se van a utilizar a lo largo de este proyecto

In [285]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

Posterior a ello, vamos a traernos la base de datos que se encuentra guardada en el repo del proyecto

In [286]:
url = 'https://raw.githubusercontent.com/omar-tueme/ciencia-de-datos-avanzada-mek-enero-2026/main/datos_info/adolescentes_ensanut2024_w.csv'
df = pd.read_csv(url, sep=';')

Ya se tienen identificadas con anterioridad las variables de interes, entonces hacemos una selección de solo esas variables y creamos un nuevo DataFrame con solo las columnas que nos interesan. El nuevo DataFrame se llama ***df_clabe***

In [287]:
columns_to_keep = ['FOLIO_I', 'FOLIO_INT', 'd0415', 'edad', 'sexo', 'd0101', 'd0107',
                   'd0108', 'd01a03a', 'd01a03b', 'd01a03c', 'd01a03d', 'd01a08a',
                   'd01a08b', 'd01a08c', 'd01a08d', 'd01a08e', 'd01a08f', 'd01a08g',
                   'd01a08h', 'd01a08i', 'd02o1a', 'd02o1b', 'd02o1c', 'd02o1d',
                   'd0313', 'd0421', 'd0422', 'd0601a', 'd0601b', 'd0601c', 'd0601d',
                   'd0601e', 'd0601f', 'd0601g', 'd0817', 'd0818', 'd06a1', 'd06a2',
                   'd06a3', 'd06a4', 'd06a5', 'd06a6', 'd06a7', 'd06a8', 'd06a9',
                   'd06a10', 'd0701', 'd0702', 'd0709', 'd0309', 'd0801', 'd0810']

df_clean = df[columns_to_keep]

Adicional a ello, ya que el nombre de columnas es cero intuitivo creamos un dicionnario que incluye el nombre de la columna, la pregunta que representa y las posibles respuestas. Este diccionario se volverá critico más adelante cuando hagamos la transformación de nuestras variables para poder hacer un correcto analisis

In [288]:
dictionary_non_id_variables = {
    'd0415': {
        'Question': 'En comparación con las personas de tu misma edad, ¿tienes dificultad para aprender cosas?',
        1: 'Ninguna dificultad',
        2: 'Cierta dificultad',
        3: 'Mucha dificultad',
        4: 'Le resulta imposible aprender',
    },
    'edad': {
        'Question': 'Edad',
        10: 10,
        11: 11,
        12: 12,
        13: 13,
        14: 14,
        15: 15,
        16: 16,
        17: 17,
        18: 18,
        19: 19,
    },
    'sexo': {
        'Question': 'Sexo',
        1: 'Hombre',
        2: 'Mujer'
    },
    'd0101': {
        'Question': 'Actualmente ¿fumas tabaco...',
        1: 'todos los días?',
        2: 'algunos días?',
        3: 'no fuma actualmente',
        9: 'No responde'
    },
    'd0107': {
        'Question': '¿Actualmente vapeas o consumes cigarros electrónicos...',
        1: 'todos los días?',
        2: 'algunos días?',
        3: 'actualmente no consume',
        4: 'No conoce el cigarro electrónico',
        9: 'No responde'
    },
    'd0108': {
        'Question': 'En los últimos 12 meses, ¿con qué frecuencia tomaste al menos una copa ya sea de vino, cerveza, whisky o cualquier otra bebida que contenga alcohol?',
        1: 'Diario',
        2: 'Semanal',
        3: 'Mensual',
        4: 'Anual',
        5: 'No he consumido en los últimos 12 meses',
        6: 'Nunca he consumido',
        9: 'No responde'
    },
    'd01a03a': {
        'Question': 'Casi todo el mundo toma medicamentos y sustancias por varias razones. Me podrías decir si has tomado, usado, probado opiáceos',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd01a03b': {
        'Question': 'Casi todo el mundo toma medicamentos y sustancias por varias razones. Me podrías decir si has tomado, usado, probado tranquilizantes',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd01a03c': {
        'Question': 'Casi todo el mundo toma medicamentos y sustancias por varias razones. Me podrías decir si has tomado, usado, probado sedantes y babitúricos',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd01a03d': {
        'Question': 'Casi todo el mundo toma medicamentos y sustancias por varias razones. Me podrías decir si has tomado, usado, probado anfetaminas o estimulantes',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd01a08a': {
        'Question': '¿Me podrías decir si has tomado, usado, probado mariguana?',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd01a08b': {
        'Question': '¿Me podrías decir si has tomado, usado, probado cocaína?',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd01a08c': {
        'Question': '¿Me podrías decir si has tomado, usado, probado crack?',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd01a08d': {
        'Question': '¿Me podrías decir si has tomado, usado, probado alucinógenos?',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd01a08e': {
        'Question': '¿Me podrías decir si has tomado, usado, probado inhalables?',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd01a08f': {
        'Question': '¿Me podrías decir si has tomado, usado, probado heroína?',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd01a08g': {
        'Question': '¿Me podrías decir si has tomado, usado, probado estimulantes tipo anfetamínico?',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd01a08h': {
        'Question': '¿Me podrías decir si has tomado, usado, probado otras drogas?',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd01a08i': {
        'Question': '¿Me podrías decir si has tomado, usado, probado mariguana sintética?',
        1: 'Sí',
        2: 'No',
        9: 'No sabe/No contesta'
    },
    'd02o1a': {
        'Question': '¿Algún médico te ha dicho que tienes diabetes o la azúcar alta en la sangre?',
        1: 'Sí',
        2: 'No',
        9: 'No sé'
    },
    'd02o1b': {
        'Question': '¿Algún médico te ha dicho que tienes la presión alta o hipertensión?',
        1: 'Sí',
        2: 'No',
        9: 'No sé'
    },
    'd02o1c': {
        'Question': '¿Algún médico te ha dicho que tienes el colesterol alto?',
        1: 'Sí',
        2: 'No',
        9: 'No sé'
    },
    'd02o1d': {
        'Question': '¿Algún médico te ha dicho que tienes los triglicéridos altos?',
        1: 'Sí',
        2: 'No',
        9: 'No sé'
    },
    'd0313': {
        'Question': '¿Alguna vez has estado embarazada?',
        1: 'Sí',
        2: 'No',
        9: 'No sé'
    },
    'd0421': {
        'Question': 'Me gustaría saber con qué frecuencia te muestras muy ansioso/a, nervioso/a o preocupado/a.',
        1: 'Diariamente',
        2: 'Semanalmente',
        3: 'Mensualmente',
        4: 'Varias veces al año',
        5: 'Nunca'
    },
    'd0422': {
        'Question': 'Me gustaría saber con qué frecuencia te muestras muy triste o deprimido/a.',
        1: 'Diariamente',
        2: 'Semanalmente',
        3: 'Mensualmente',
        4: 'Varias veces al año',
        5: 'Nunca'
    },
    'd0601a': {
        'Question': 'Durante la última semana... ¿Sentías como si no pudieras quitarte de encima la tristeza?',
        1: 'Rara vez o nunca (menos de un día)',
        2: 'Pocas veces o alguna vez (1-2 días)',
        3: 'Un número de veces considerables (3-4 días)',
        4: 'Todo el tiempo o la mayoria del tiempo (5-7 días)'
    },
    'd0601b': {
        'Question': 'Durante la última semana... ¿Te costaba concentrarte en lo que estabas haciendo?',
        1: 'Rara vez o nunca (menos de un día)',
        2: 'Pocas veces o alguna vez (1-2 días)',
        3: 'Un número de veces considerables (3-4 días)',
        4: 'Todo el tiempo o la mayoria del tiempo (5-7 días)'
    },
    'd0601c': {
        'Question': 'Durante la última semana... ¿Te sentiste deprimido/a?',
        1: 'Rara vez o nunca (menos de un día)',
        2: 'Pocas veces o alguna vez (1-2 días)',
        3: 'Un número de veces considerables (3-4 días)',
        4: 'Todo el tiempo o la mayoria del tiempo (5-7 días)'
    },
    'd0601d': {
        'Question': 'Durante la última semana... ¿Te parecía que todo lo que hacías era un esfuerzo?',
        1: 'Rara vez o nunca (menos de un día)',
        2: 'Pocas veces o alguna vez (1-2 días)',
        3: 'Un número de veces considerables (3-4 días)',
        4: 'Todo el tiempo o la mayoria del tiempo (5-7 días)'
    },
    'd0601e': {
        'Question': 'Durante la última semana... ¿No dormiste bien?',
        1: 'Rara vez o nunca (menos de un día)',
        2: 'Pocas veces o alguna vez (1-2 días)',
        3: 'Un número de veces considerables (3-4 días)',
        4: 'Todo el tiempo o la mayoria del tiempo (5-7 días)'
    },
    'd0601f': {
        'Question': 'Durante la última semana... ¿Disfrutaste de la vida?',
        1: 'Rara vez o nunca (menos de un día)',
        2: 'Pocas veces o alguna vez (1-2 días)',
        3: 'Un número de veces considerables (3-4 días)',
        4: 'Todo el tiempo o la mayoria del tiempo (5-7 días)'
    },
    'd0601g': {
        'Question': 'Durante la última semana... ¿Te sentiste triste?',
        1: 'Rara vez o nunca (menos de un día)',
        2: 'Pocas veces o alguna vez (1-2 días)',
        3: 'Un número de veces considerables (3-4 días)',
        4: 'Todo el tiempo o la mayoria del tiempo (5-7 días)'
    },
    'd0817': {
        'Question': '¿Alguna vez has pensado en suicidarte?',
        1: 'Sí',
        2: 'No',
        8: 'No responde'
    },
    'd0818': {
        'Question': '¿Con qué frecuencia lo has pensado?',
        1: 'Siempre o muy frecuente',
        2: 'Casi siempre',
        3: 'Algunas veces',
        4: 'Rara vez',
        99: 'No responde'
    },
    'd06a1': {
        'Question': '¿Con qué frecuencia en los últimos 3 meses… te ha preocupado engordar?',
        1: 'Nunca o casi nunca',
        2: 'Algunas veces',
        3: 'Frecuentemente (2 veces en una semana)',
        4: 'Muy frecuentemente (más de 2 veces en una semana)'
    },
    'd06a2': {
        'Question': '¿Con qué frecuencia en los últimos 3 meses… en ocasiones, has comido demasiado?',
        1: 'Nunca o casi nunca',
        2: 'Algunas veces',
        3: 'Frecuentemente (2 veces en una semana)',
        4: 'Muy frecuentemente (más de 2 veces en una semana)'
    },
    'd06a3': {
        'Question': '¿Con qué frecuencia en los últimos 3 meses… has perdido el control sobre lo que comes?',
        1: 'Nunca o casi nunca',
        2: 'Algunas veces',
        3: 'Frecuentemente (2 veces en una semana)',
        4: 'Muy frecuentemente (más de 2 veces en una semana)'
    },
    'd06a4': {
        'Question': '¿Con qué frecuencia en los últimos 3 meses… has vomitado después de comer para bajar de peso?',
        1: 'Nunca o casi nunca',
        2: 'Algunas veces',
        3: 'Frecuentemente (2 veces en una semana)',
        4: 'Muy frecuentemente (más de 2 veces en una semana)'
    },
    'd06a5': {
        'Question': '¿Con qué frecuencia en los últimos 3 meses… has hecho ayunos (dejar de comer por 12 horas o más) para tratar de bajar de peso?',
        1: 'Nunca o casi nunca',
        2: 'Algunas veces',
        3: 'Frecuentemente (2 veces en una semana)',
        4: 'Muy frecuentemente (más de 2 veces en una semana)'
    },
    'd06a6': {
        'Question': '¿Con qué frecuencia en los últimos 3 meses… has hecho dietas para tratar de bajar de peso?',
        1: 'Nunca o casi nunca',
        2: 'Algunas veces',
        3: 'Frecuentemente (2 veces en una semana)',
        4: 'Muy frecuentemente (más de 2 veces en una semana)'
    },
    'd06a7': {
        'Question': '¿Con qué frecuencia en los últimos 3 meses… has hecho ejercicio en exceso para tratar de bajar de peso?',
        1: 'Nunca o casi nunca',
        2: 'Algunas veces',
        3: 'Frecuentemente (2 veces en una semana)',
        4: 'Muy frecuentemente (más de 2 veces en una semana)'
    },
    'd06a8': {
        'Question': '¿Con qué frecuencia en los últimos 3 meses… has usado pastillas para tratar de bajar de peso?',
        1: 'Nunca o casi nunca',
        2: 'Algunas veces',
        3: 'Frecuentemente (2 veces en una semana)',
        4: 'Muy frecuentemente (más de 2 veces en una semana)'
    },
    'd06a9': {
        'Question': '¿Con qué frecuencia en los últimos 3 meses… has tomado diuréticos (sustancia para perder agua) para tratar de bajar de peso?',
        1: 'Nunca o casi nunca',
        2: 'Algunas veces',
        3: 'Frecuentemente (2 veces en una semana)',
        4: 'Muy frecuentemente (más de 2 veces en una semana)'
    },
    'd06a10': {
        'Question': '¿Con qué frecuencia en los últimos 3 meses… has tomado laxantes (sustancia para facilitar la evacuación) para tratar de bajar de peso?',
        1: 'Nunca o casi nunca',
        2: 'Algunas veces',
        3: 'Frecuentemente (2 veces en una semana)',
        4: 'Muy frecuentemente (más de 2 veces en una semana)'
    },
    'd0701': {
        'Question': '¿Sufriste algún daño a tu salud a causa de un accidente en los últimos 12 meses?',
        1: 'Sí',
        2: 'No',
        8: 'No sabe',
        9: 'No responde'
    },
    'd0702': {
        'Question': '¿Cómo fue que te accidentaste?',
        1: 'Choque entre vehículos de 4 o más ruedas',
        2: 'Choque en motocicleta',
        3: 'Choque o atropellamiento como ciclista',
        4: 'Atropellado como peatón',
        5: 'Otros accidentes de transporte: avión, tren, lancha',
        6: 'Caída a nivel de piso',
        7: 'Golpe con objeto, equipo o maquinaria',
        8: 'Caída de un nivel a otro: silla, escalera, etc.',
        9: 'Golpe, rasguño o mordida de persona o animal (caballo, perro, etc.)',
        10: 'Exposición a corriente eléctrica, radiación',
        11: 'Exposición a humo o fuego',
        12: 'Contacto con superficies o sustancias calientes (quemaduras)',
        13: 'Exposición a fuerzas de la naturaleza: rayo, granizo',
        14: 'Picadura de animal ponzoñoso o contacto con plantas venenosas',
        15: 'Envenenamiento accidental: alimentos, medicamentos',
        16: 'Sumersión en depósitos de agua: alberca, cisterna',
        17: 'Otros accidentes que obstruyen la respiración: objeto extraño, alimento',
        18: 'Exposición accidental a otros factores',
        99: 'NS/NR'
    },
    'd0709': {
        'Question': '¿Qué problema de salud permanente te ocasionó el accidente?',
        1: 'Ningún problema',
        2: 'Limitación o dificultad para moverse o caminar (o lo hace con ayuda)',
        3: 'Limitación o dificultad para usar sus brazos y manos',
        4: 'Dificultad para oír',
        5: 'Dificultad para hablar',
        6: 'Dificultad para ver o ceguera',
        7: 'Otra limitación física o mental',
        99: 'NS/NR'
    },
    'd0309': {
        'Question': '¿Fuiste presionada(o) o forzada(o) para tener esa primera relación sexual?',
        1: 'Sí',
        2: 'No',
        9: 'No recuerdo'
    },
    'd0801': {
        'Question': 'En los últimos 12 meses, ¿sufriste algún daño a tu salud por robo, agresión o violencia?',
        1: 'Sí',
        2: 'No',
        9: 'No responde'
    },
    'd0810': {
        'Question': '¿A lo largo de tu vida, alguien te manoseó, tocó o acarició alguna parte de tu cuerpo o tuvo relaciones sexuales contigo cuando eras muy pequeño/a?',
        1: 'Sí, antes de los 12',
        2: 'Sí, cuando tenía 12 o más años',
        3: 'No, nunca',
        4: 'No recuerda',
        5: 'No responde'
    }
}

Empezamos con información general de DataFrame, se van a hacer varias transformaciones pero esto nos sirve para ver como está construido el DataFrame y que tenemos un total de 53 columnas y 3,730 filas. Cuando nos metemos a estudiar cada una de las variables, encontramos algunas concideraciones importantes (otra vez, esto es a gran escala, más adelante nos enfocaremos en entender cada una de nuestras variables)

* Los "nulls" se encuentrán guardados como " " por lo que los transformaremos más adelante en nulls. Éstas columnas con " " son de tipo objecto

* La variable objetivo ***d0415*** (En comparación con las personas de tu misma edad, ¿tienes dificultad para aprender cosas?) tiene 690 nulls, estas observaciones se tendrán que quitar del análisis ya que no aportan nada

* Hay preguntas como ***d0313*** (¿Alguna vez has estado embarazada?), ***d0818*** (¿Con qué frecuencia lo has pensado [suicidarte]?), ***d0702*** (¿Cómo fue que te accidentaste?), ***d0709*** (¿Qué problema de salud permanente te ocasionó el accidente?) y ***d0309*** (¿Fuiste presionada(o) o forzada(o) para tener esa primera relación sexual?) con una cantidad muy elevada de valores faltantes (por la misma naturaleza de la pregunta). Se evaluara con precación si dichas columan deberían de ser excluidas del análisis


In [289]:
df_clean.head()

,FOLIO_I,FOLIO_INT,d0415,edad,sexo,d0101,d0107,d0108,d01a03a,d01a03b,...,d06a7,d06a8,d06a9,d06a10,d0701,d0702,d0709,d0309,d0801,d0810
0,2024_01001001,2024_01001001_05,1,11,2,3,4,6,,,...,1,1,1,1,2,,,,2,3
1,2024_01001006,2024_01001006_03,,19,2,3,3,3,2,2,...,1,1,1,1,2,,,2,2,1
2,2024_01001009,2024_01001009_03,1,15,1,3,3,2,2,2,...,1,1,1,1,2,,,,2,9
3,2024_01001024,2024_01001024_03,,19,1,2,3,2,2,2,...,1,1,1,1,2,,,,2,3
4,2024_01001026,2024_01001026_03,,19,1,3,2,3,2,2,...,1,1,1,1,2,,,2,2,3


In [290]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3730 entries, 0 to 3729
Data columns (total 53 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   FOLIO_I    3730 non-null   object
 1   FOLIO_INT  3730 non-null   object
 2   d0415      3730 non-null   object
 3   edad       3730 non-null   int64 
 4   sexo       3730 non-null   int64 
 5   d0101      3730 non-null   int64 
 6   d0107      3730 non-null   int64 
 7   d0108      3730 non-null   int64 
 8   d01a03a    3730 non-null   object
 9   d01a03b    3730 non-null   object
 10  d01a03c    3730 non-null   object
 11  d01a03d    3730 non-null   object
 12  d01a08a    3730 non-null   object
 13  d01a08b    3730 non-null   object
 14  d01a08c    3730 non-null   object
 15  d01a08d    3730 non-null   object
 16  d01a08e    3730 non-null   object
 17  d01a08f    3730 non-null   object
 18  d01a08g    3730 non-null   object
 19  d01a08h    3730 non-null   object
 20  d01a08i    3730 non-null   obj

In [291]:
for col in df_clean.columns:
    print(f'Variable: {col}.\nNumber unique values: {df_clean[col].nunique()}.\n{df_clean[col].value_counts()}\n\n')

Variable: FOLIO_I.
Number unique values: 3730.
FOLIO_I
2024_32056030    1
2024_01001001    1
2024_32052016    1
2024_32052015    1
2024_32052014    1
                ..
2024_01001032    1
2024_01001028    1
2024_01001026    1
2024_01001024    1
2024_01001009    1
Name: count, Length: 3730, dtype: int64


Variable: FOLIO_INT.
Number unique values: 3730.
FOLIO_INT
2024_32056030_03    1
2024_01001001_05    1
2024_32052016_04    1
2024_32052015_05    1
2024_32052014_04    1
                   ..
2024_01001032_02    1
2024_01001028_04    1
2024_01001026_03    1
2024_01001024_03    1
2024_01001009_03    1
Name: count, Length: 3730, dtype: int64


Variable: d0415.
Number unique values: 5.
d0415
1    2461
      690
2     532
3      46
4       1
Name: count, dtype: int64


Variable: edad.
Number unique values: 10.
edad
11    423
13    421
15    398
10    387
12    378
17    366
18    345
19    344
14    335
16    333
Name: count, dtype: int64


Variable: sexo.
Number unique values: 2.
sexo
1   

## Limpieza de la base de datos

Empezaremos categorizar todos los " " como nulls

Después de hacerlo nos damos cuenta que si tenemos nulls, no como se vió en el primer df_clean.info

Vamos a aprovechar para remover las columnas de folio ya que no sirven para el análisis

Por último se va a convertir todas las columnas a Int64 ya que en realidad lo que se tienen son números y no texto

In [292]:
df_clean = df_clean.replace(' ', np.nan)

In [293]:
df_clean = df_clean.drop(columns=['FOLIO_I', 'FOLIO_INT'])

df_clean = df_clean.astype('Int64')

In [294]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3730 entries, 0 to 3729
Data columns (total 51 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   d0415    3040 non-null   Int64
 1   edad     3730 non-null   Int64
 2   sexo     3730 non-null   Int64
 3   d0101    3730 non-null   Int64
 4   d0107    3730 non-null   Int64
 5   d0108    3730 non-null   Int64
 6   d01a03a  2927 non-null   Int64
 7   d01a03b  2927 non-null   Int64
 8   d01a03c  2927 non-null   Int64
 9   d01a03d  2927 non-null   Int64
 10  d01a08a  2927 non-null   Int64
 11  d01a08b  2927 non-null   Int64
 12  d01a08c  2927 non-null   Int64
 13  d01a08d  2927 non-null   Int64
 14  d01a08e  2927 non-null   Int64
 15  d01a08f  2927 non-null   Int64
 16  d01a08g  2927 non-null   Int64
 17  d01a08h  2927 non-null   Int64
 18  d01a08i  2927 non-null   Int64
 19  d02o1a   3730 non-null   Int64
 20  d02o1b   3730 non-null   Int64
 21  d02o1c   3730 non-null   Int64
 22  d02o1d   3730 non-null  

Posterior, se van a remover todas las observaciones que no tienen un valor en nuestra variable objetivo. Esto porque dichas observaciones no estarían aportando a este análisis en especifico. Esto la vamos a guardar en un nuevo DataFrame llamado ***df_clean1***

Ésto nos deja con un total de 53 columnas y 3040 renglonse

In [295]:
df_clean1 = df_clean.dropna(subset=['d0415']).reset_index(drop=True)

df_clean1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3040 entries, 0 to 3039
Data columns (total 51 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   d0415    3040 non-null   Int64
 1   edad     3040 non-null   Int64
 2   sexo     3040 non-null   Int64
 3   d0101    3040 non-null   Int64
 4   d0107    3040 non-null   Int64
 5   d0108    3040 non-null   Int64
 6   d01a03a  2237 non-null   Int64
 7   d01a03b  2237 non-null   Int64
 8   d01a03c  2237 non-null   Int64
 9   d01a03d  2237 non-null   Int64
 10  d01a08a  2237 non-null   Int64
 11  d01a08b  2237 non-null   Int64
 12  d01a08c  2237 non-null   Int64
 13  d01a08d  2237 non-null   Int64
 14  d01a08e  2237 non-null   Int64
 15  d01a08f  2237 non-null   Int64
 16  d01a08g  2237 non-null   Int64
 17  d01a08h  2237 non-null   Int64
 18  d01a08i  2237 non-null   Int64
 19  d02o1a   3040 non-null   Int64
 20  d02o1b   3040 non-null   Int64
 21  d02o1c   3040 non-null   Int64
 22  d02o1d   3040 non-null  

Por último, para que el proceso de limpieza sea mucho más sencillo se va a sustituir los valores de las celdas y los títulos de las columnas con lo que se tiene en el diccionario. Esto es meramente transitorio para poder hacer una limpieza de datos más sencilla

In [296]:
"""
NOT USED -- TO CHANGE COLUMN NAMES BUT IT MAKES SUBSEQUENT CLEANING VERY DIFFICULT

rename_dict = {
    var: info["Question"]
    for var, info in dictionary_non_id_variables.items()
    if var in df_clean1.columns
}

df_clean1 = df_clean1.rename(columns=rename_dict)
"""

'\nNOT USED -- TO CHANGE COLUMN NAMES BUT IT MAKES SUBSEQUENT CLEANING VERY DIFFICULT\n\nrename_dict = {\n    var: info["Question"]\n    for var, info in dictionary_non_id_variables.items()\n    if var in df_clean1.columns\n}\n\ndf_clean1 = df_clean1.rename(columns=rename_dict)\n'

In [297]:
for var, info in dictionary_non_id_variables.items():
    if var in df_clean1.columns:
        mapping = {k: v for k, v in info.items() if isinstance(k, int)}

        df_clean1[var] = df_clean1[var].map(mapping)

In [298]:
df_clean1

,d0415,edad,sexo,d0101,d0107,d0108,d01a03a,d01a03b,d01a03c,d01a03d,...,d06a7,d06a8,d06a9,d06a10,d0701,d0702,d0709,d0309,d0801,d0810
0,Ninguna dificultad,11,Mujer,no fuma actualmente,No conoce el cigarro electrónico,Nunca he consumido,NaN,NaN,NaN,NaN,...,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,No,NaN,NaN,NaN,No,"No, nunca"
1,Ninguna dificultad,15,Hombre,no fuma actualmente,actualmente no consume,Semanal,No,No,No,No,...,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,No,NaN,NaN,NaN,No,NaN
2,Ninguna dificultad,13,Hombre,no fuma actualmente,actualmente no consume,Nunca he consumido,No,No,No,No,...,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,No,NaN,NaN,NaN,No,"No, nunca"
3,Ninguna dificultad,10,Mujer,no fuma actualmente,No conoce el cigarro electrónico,Nunca he consumido,NaN,NaN,NaN,NaN,...,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,No,NaN,NaN,NaN,No,"No, nunca"
4,Ninguna dificultad,13,Mujer,no fuma actualmente,actualmente no consume,Nunca he consumido,No,No,No,No,...,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,No,NaN,NaN,NaN,No,"No, nunca"
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3035,Ninguna dificultad,13,Hombre,no fuma actualmente,actualmente no consume,Nunca he consumido,No,No,No,No,...,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,No,NaN,NaN,NaN,No,"No, nunca"
3036,Ninguna dificultad,13,Mujer,no fuma actualmente,actualmente no consume,Nunca he consumido,No,No,No,No,...,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,No,NaN,NaN,NaN,No,"No, nunca"
3037,Ninguna dificultad,11,Mujer,no fuma actualmente,actualmente no consume,No he consumido en los últimos 12 meses,NaN,NaN,NaN,NaN,...,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,No,NaN,NaN,NaN,No,"No, nunca"
3038,Ninguna dificultad,11,Mujer,no fuma actualmente,actualmente no consume,Nunca he consumido,NaN,NaN,NaN,NaN,...,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,Nunca o casi nunca,No,NaN,NaN,NaN,No,"No, nunca"


Con la base de datos más entendible, vamos a proceder a enfocarnos en cada una de las subsecciones y en realizar la limpieza y preparación pertinente. La intensión es convertir la mayoría de las variables en variables dummies para poder hacer un análisis categórico

### Variable dependiente

Empezemos por recordar cómo está construida la variable dependiente. 2,461 de 3,040 (80%) dice no tener ninguna dificultad aprendiendo en comparación a personas de su misma edad

Dado el bajo volumen en las respuestas de "mucha dificultad" y "le resulta imposible aprender", se van a agrupar con la respuesta" cierta dificultad" para construir la variable dicotómica que compara ninguna dificultad de aprender versus alguna dificultad para aprender

Todas estas transformación se irán poniendo en un nuevo DataFrame llamado ***df_final*** que es el que se utilizará para realizar el análisis

In [299]:
df_clean1['d0415'].value_counts()

,count
d0415,
Ninguna dificultad,2461
Cierta dificultad,532
Mucha dificultad,46
Le resulta imposible aprender,1


In [300]:
df_final = pd.DataFrame()

df_final['Dificultad para aprender'] = (df_clean1['d0415'] != 'Ninguna dificultad').astype(int)

df_final['Dificultad para aprender'].value_counts()

,count
Dificultad para aprender,
0,2461
1,579


In [301]:
df_final

,Dificultad para aprender
0,0
1,0
2,0
3,0
4,0
...,...
3035,0
3036,0
3037,0
3038,0


### Variables de identidad

Posterior a la variable dependiente, tenemos las variables de identidad (sexo y edad), recordemos su distribución. De ahora en adelante siempre se correra crosstab contra la variable dependiendo para ir observando patrones

Parece ser que al limpiar por nuestra variable dependiente se impacto desproporcionalmente a los jovenes de edad entre los 18 y 19 años. Al revisar la encuesta, existe una leyenda que indica al entrevistador pasarse la pregunta de interes para las personas de esa edad. Es por ello que, por consistencia en los resultados, se eliminarán los 6 registros de las personas con edad entre 18 y 19 años -- dicha limpieza se hará hasta el final de armar df_final para evitar que los datos no empaten al ir construyendo el DataFrame de manera correcta con las siguientes subsecciones

Respecto al sexo, se mantiene una distribución muy cercana a 50/50, por lo que se mantiene la variable como está. Ambas variables se guardaran en ***df_final***, edad siendo una variable discreta y sexo siendo una variable dicotómica

In [302]:
df_clean1['edad'].value_counts().sort_index()

,count
edad,
10,387
11,423
12,378
13,421
14,335
15,397
16,333
17,360
18,5


In [303]:
df_clean['edad'].value_counts().sort_index()

,count
edad,
10,387
11,423
12,378
13,421
14,335
15,398
16,333
17,366
18,345


In [304]:
pd.crosstab(
    df_clean1['edad'],
    df_clean1['d0415'],
)

d0415,Cierta dificultad,Le resulta imposible aprender,Mucha dificultad,Ninguna dificultad
edad,,,,
10,88,0,7,292
11,89,1,4,329
12,55,0,12,311
13,94,0,6,321
14,56,0,5,274
15,52,0,7,338
16,46,0,3,284
17,52,0,2,306
18,0,0,0,5


In [305]:
df_clean1['sexo'].value_counts().sort_index()

,count
sexo,
Hombre,1541
Mujer,1499


In [306]:
pd.crosstab(
    df_clean1['sexo'],
    df_clean1['d0415'],
)

d0415,Cierta dificultad,Le resulta imposible aprender,Mucha dificultad,Ninguna dificultad
sexo,,,,
Hombre,261,0,22,1258
Mujer,271,1,24,1203


In [307]:
df_final['Edad'] = df_clean1['edad']

df_final['Edad'].value_counts().sort_index()

,count
Edad,
10,387
11,423
12,378
13,421
14,335
15,397
16,333
17,360
18,5


In [308]:
df_final['Es mujer'] = (df_clean1['sexo'] == 'Mujer').astype(int)

df_final['Es mujer'].value_counts()

,count
Es mujer,
0,1541
1,1499


In [309]:
df_final

,Dificultad para aprender,Edad,Es mujer
0,0,11,1
1,0,15,0
2,0,13,0
3,0,10,1
4,0,13,1
...,...,...,...
3035,0,13,0
3036,0,13,1
3037,0,11,1
3038,0,11,1


### Variables de drogas - Tabaco

Recordemos como se distribuyen estas variables. Dado lo que se vió arriba con la variable edad en donde parece que hay preguntas que no se hace a cierta edad, vamos a correr crosstab contra la variable objetivo y contra edad para ser precavidos en todos los futuros chequeos

Parecen haber pocos casos de jovenes que fuman (buenas noticias!), para generar la variable dummy se harán 3 agrupaciones

* "No responde" y "No conoce el cigarro electrónico" se le asignará la etiqueta de no fuma actualmente dado que esta respuesta es la moda

* "algunos días?" y "todos los días?" se agruparan para representar que el joven fuma (sin importar la frecuencia)

* Para finalizar, ambas columnas se colapsaran a 1 sola, si una persona fuma o consume cigarro electronico se considerará fumador. Si no consume ninguno de los dos, se considerará un no fumador

In [310]:
df_clean1['d0101'].value_counts().sort_index()

,count
d0101,
No responde,38
algunos días?,43
no fuma actualmente,2949
todos los días?,10


In [311]:
pd.crosstab(
    df_clean1['d0101'],
    df_clean1['d0415'],
)

d0415,Cierta dificultad,Le resulta imposible aprender,Mucha dificultad,Ninguna dificultad
d0101,,,,
No responde,7,0,0,31
algunos días?,5,0,0,38
no fuma actualmente,517,1,46,2385
todos los días?,3,0,0,7


In [312]:
pd.crosstab(
    df_clean1['d0101'],
    df_clean1['edad'],
)

edad,10,11,12,13,14,15,16,17,18,19
d0101,,,,,,,,,,
No responde,3,7,3,9,3,2,7,4,0,0
algunos días?,1,0,3,1,4,7,10,17,0,0
no fuma actualmente,383,416,372,410,327,386,315,334,5,1
todos los días?,0,0,0,1,1,2,1,5,0,0


In [316]:
df_clean1['d0107'].value_counts().sort_index()

,count
d0107,
No conoce el cigarro electrónico,1084
No responde,15
actualmente no consume,1875
algunos días?,64
todos los días?,2


In [317]:
pd.crosstab(
    df_clean1['d0107'],
    df_clean1['d0415'],
)

d0415,Cierta dificultad,Le resulta imposible aprender,Mucha dificultad,Ninguna dificultad
d0107,,,,
No conoce el cigarro electrónico,198,1,24,861
No responde,3,0,0,12
actualmente no consume,317,0,22,1536
algunos días?,14,0,0,50
todos los días?,0,0,0,2


In [318]:
pd.crosstab(
    df_clean1['d0107'],
    df_clean1['edad'],
)

edad,10,11,12,13,14,15,16,17,18,19
d0107,,,,,,,,,,
No conoce el cigarro electrónico,229,223,147,127,101,102,75,76,3,1
No responde,0,4,0,1,1,1,5,3,0,0
actualmente no consume,157,195,227,282,224,286,240,262,2,0
algunos días?,1,1,4,11,9,8,13,17,0,0
todos los días?,0,0,0,0,0,0,0,2,0,0


In [320]:
pd.crosstab(
    df_clean1['d0107'],
    df_clean1['d0101'],
)

d0101,No responde,algunos días?,no fuma actualmente,todos los días?
d0107,,,,
No conoce el cigarro electrónico,11,8,1063,2
No responde,11,0,4,0
actualmente no consume,15,24,1830,6
algunos días?,1,10,51,2
todos los días?,0,1,1,0


In [319]:
df_final['Fuma'] = ((df_clean1['d0101'] == 'algunos días?') | (df_clean1['d0101'] == 'todos los días?') | (df_clean1['d0107'] == 'algunos días?') | (df_clean1['d0101'] == 'todos los días?')).astype(int)

df_final['Fuma'].value_counts()

,count
Fuma,
0,2935
1,105


In [321]:
df_final

,Dificultad para aprender,Edad,Es mujer,Fuma
0,0,11,1,0
1,0,15,0,0
2,0,13,0,0
3,0,10,1,0
4,0,13,1,0
...,...,...,...,...
3035,0,13,0,0
3036,0,13,1,0
3037,0,11,1,0
3038,0,11,1,0


### Variables de drogas - Alcohol

Recordemos como se distribuye esta variable. Dado que hay riqueza adicional en esta respuesta respecto a la frecuencia de consumo, no es de nuestro interes convertirla en una variable dictómica. Lo que se propone hacer es volvera una variable continua entre 0 y 1 en donde 0 es consumir alcohol 0 días al año y 1 es consumir alcohol 365 días al año. De este modo, se tendrán los siguientes valores =

* Diario = 1

* Semanal = 52 / 365 = 0.142466

* Mensual = 12 / 365 = 0.032977

* Anual = 1 / 365 = 0.002740

* No he consumido en los últimos 12 meses + No responde + Nunca he consumido = 0

Es cierto, sin embargo, que esta distribución no es la óptima, pero consideramos que es mejor guardar esta variable de manera continua en vez de guardarla de manera dicotómica


In [322]:
df_clean1['d0108'].value_counts().sort_index()

,count
d0108,
Anual,231
Diario,3
Mensual,96
No he consumido en los últimos 12 meses,660
No responde,7
Nunca he consumido,2001
Semanal,42


In [323]:
pd.crosstab(
    df_clean1['d0108'],
    df_clean1['d0415'],
)

d0415,Cierta dificultad,Le resulta imposible aprender,Mucha dificultad,Ninguna dificultad
d0108,,,,
Anual,44,0,0,187
Diario,0,0,0,3
Mensual,11,0,1,84
No he consumido en los últimos 12 meses,107,0,11,542
No responde,0,0,1,6
Nunca he consumido,364,1,32,1604
Semanal,6,0,1,35


In [324]:
pd.crosstab(
    df_clean1['d0108'],
    df_clean1['edad'],
)

edad,10,11,12,13,14,15,16,17,18,19
d0108,,,,,,,,,,
Anual,2,8,6,23,27,52,47,65,1,0
Diario,0,0,0,0,0,1,1,1,0,0
Mensual,0,1,3,2,7,19,31,33,0,0
No he consumido en los últimos 12 meses,84,89,90,106,71,81,68,70,1,0
No responde,0,2,0,1,1,0,2,1,0,0
Nunca he consumido,300,323,278,286,227,238,173,172,3,1
Semanal,1,0,1,3,2,6,11,18,0,0


In [328]:
def scale_alcohol(row):
    x = row['d0108']
    if x == 'Diario':
        return 1
    elif x == 'Semanal':
        return 52/365
    elif x == 'Mensual':
        return 12/365
    elif x == 'Anual':
        return 1/365
    else:
        return 0

df_final['Alcohol'] = df_clean1.apply(scale_alcohol, axis=1)

df_final['Alcohol'].value_counts()

,count
Alcohol,
0.000000,2668
0.002740,231
0.032877,96
0.142466,42
1.000000,3


In [329]:
df_final

,Dificultad para aprender,Edad,Es mujer,Fuma,Alcohol
0,0,11,1,0,0.000000
1,0,15,0,0,0.142466
2,0,13,0,0,0.000000
3,0,10,1,0,0.000000
4,0,13,1,0,0.000000
...,...,...,...,...,...
3035,0,13,0,0,0.000000
3036,0,13,1,0,0.000000
3037,0,11,1,0,0.000000
3038,0,11,1,0,0.000000


#A
#A
#A
#A
#A
#A
#A
#A
#A
#A

In [315]:
AL FINAL, NO CORRER HASTA QUE ESTÉ DF_FINAL LISTO

df_final = df_final[~(df_final['Edad'] == 18) & ~(df_final['Edad'] == 19)]

SyntaxError: invalid syntax (ipython-input-4088970218.py, line 1)